crypto/o_str.c

```diff
@@ -90,6 +90,69 @@ size_t OPENSSL_strlcat(char *dst, const char *src, size_t size)
     return l + OPENSSL_strlcpy(dst, src, size);
 }
 
+/**
+ * @brief Converts a string to an unsigned long integer.
+ *
+ * This function attempts to convert a string representation of a number
+ * to an unsigned long integer, given a specified base. It also provides
+ * error checking and reports whether the conversion was successful.
+ * This function is just a wrapper around the POSIX strtoul function with
+ * additional error checking.  This implies that errno for the caller is set
+ * on calls to this function.
+ *
+ * @param str The string containing the representation of the number.
+ * @param endptr A pointer to a pointer to character. If not NULL, it is set
+ *               to the character immediately following the number in the
+ *               string.
+ * @param base The base to use for the conversion, which must be between 2,
+ *             and 36 inclusive, or be the special value 0. If the base is 0,
+ *             the actual base is determined by the format of the initial
+ *             characters of the string.
+ * @param num A pointer to an unsigned long where the result of the
+ *            conversion is stored.
+ *
+ * @return 1 if the conversion was successful, 0 otherwise. Conversion is
+ *         considered unsuccessful if no digits were consumed or if an error
+ *         occurred during conversion.
+ *
+ * @note It is the caller's responsibility to check if the conversion is
+ *       correct based on the expected consumption of the string as reported
+ *       by endptr.
+ */
+int OPENSSL_strtoul(const char *str, char **endptr, int base,
+                    unsigned long *num)
+{
+    char *tmp_endptr;
+    char **internal_endptr = endptr == NULL ? &tmp_endptr : endptr;
+
+    errno = 0;
+
+    *internal_endptr = (char *)str;
```

Comment:
This is not required. When you cast away a const qualifier, as in the example (char *)str, you're telling the compiler to ignore the const attribute of the original pointer. This can be problematic because it allows the code to modify the original data, even if it was intended to be constant.

Coverity will likely report a warning or error in this case, as it detects a potential issue.

Reply:
if you don't do the cast you get gcc warnings about discarding the const qualifier, breaking --strict-warnings builds

coverity shouldn't warn about anything here, as modifications to internal_endptr won't modify the target data, if something in glibc modifies **internal_endptr, that would be a valid error, but nothing actually does that

Reply:
It's odd. As I noted in another comment, we can eliminate the 'end_ptr' argument to simplify this.

Reply:
```
crypto/o_str.c: In function 'OPENSSL_strtoul':
crypto/o_str.c:130:22: error: assignment discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
  130 |     *internal_endptr = str;
      |    
```
```
nhorman@hmsbeagle:~/git/openssl$ git diff
diff --git a/crypto/o_str.c b/crypto/o_str.c
index e75e44129e..b6bea7183c 100644
--- a/crypto/o_str.c
+++ b/crypto/o_str.c
@@ -127,7 +127,7 @@ int OPENSSL_strtoul(const char *str, char **endptr, int base,
 
     errno = 0;
 
-    *internal_endptr = (char *)str;
+    *internal_endptr = str;
 
     if (num == NULL)
         return 0;
nhorman@hmsbeagle:~/git/openssl$ 
```
And we can't eliminate the end_ptr as discussed above, as there are legitimate use cases for which end_ptr won't show that the entire string was consumed.

Reply:
Agree, the cast seems to be the best way to handle this, and it is what is done by the couple of implementations of `strtoul()` I looked at